### DISCUSSION TITLE
New feature ideas

### DISCUSSION THREAD

**Discussion by yakitatepan** [PetrIvan/chord-seq-ai-app#37] [Feature requests]
Hi, first of all, thanks for this tool! I want to request a couple of ideas I think could be taken into consideration for future development.

1. Ability to save a chord progression within the desktop app and not only in your personal computer. 
2. Having a mixture of playback sounds (not just piano) when previewing chords like synths, plucks, etc.
3. When clicking a chord for the second time, the text: "Select a chord to see suggestions based on context." is previewed below. What about moving this text (and making it clickable) next to the actual chord?
4. When clicking a chord, having a piano roll show up immediately and show how the chord is played instead of having to click "Open chord variants" and seeing how to play it in another preview. 
5. Back to #4, still have the "Open chord variants" but this could be added next to the piano roll for users to decide if they want to see variants.

**Comment by yakitatepan** [comment]
Not sure why SEO optimization is added to my 5th tip but I meant going back to my (#4) tip lol.

**Comment by PetrIvan** [comment]
Hi @yakitatepan, thank you for the feedback! I'm glad that you find ChordSeqAI useful!

I understand that saving within the app could be useful, however, I am worried that users might lose their progressions. Browser storage is not a good way to work with potentially valuable data (as it can accidentally get deleted), and setting up a cloud solution or server-side storage is currently beyond what I can afford.

I appreciate that you brought up the playback update! It seems interesting to me and I already have some improvements planned, in fact, there is [a feature request](https://github.com/PetrIvan/chord-seq-ai-app/discussions/19) that I've created briefly mentioning it.

The text is just a tip for new users who might get stuck and don't know how to get the suggestions, it is shown by default when there aren't any chords selected. Moving the text to the timeline would make it a bit cluttered in my opinion.

Regarding points 4 and 5, chord variants were originally included as an afterthought to solve the issue of different notations for essentially the same chords. I see a chord as a fundamental block in the sense that when I see its notation, the exact notes and their octave are not that important to me. The focus on a chord as a building block is fairly common among musicians because it helps to convey the message regardless of the instrument or its voicing. Although now that I think about it, some visualization other than the notation could be useful, let me know if you have some ideas of how it could be implemented while making the layout nice and clean!

**Reply by yakitatepan** [reply]
Hey! thanks for your reply. Let me know if you need help for future updates like adding different sounds. :) In the meantime, if anything comes to me as an idea (for layout) I will let you know!

**Comment by abulka** [comment]
Drag and Drop to rearrange the chord progressions is a much needed feature.

Also, there is a massive UI bug - after zooming in and out the beginning of the song disappears and I can't get back to viewing the chords - the chords are always out of view. In other words basic sideways scrolling seems to be missing (Chrome, Mac, Trackpad).

**Reply by PetrIvan** [reply]
Thanks, I'll consider adding it along with mobile support.

Zooming in/out is calculated according to the position of the cursor, so if the cursor stays in place, the timeline shouldn't shift. However, since the progression cannot be moved beyond the first chord, it gets clamped and potentially moves the view anyway. You can use this to zoom in on the beginning and zoom out on the end, which moves the view to the front (just like in any other editor you may be familiar with).

For easier navigation, especially on a trackpad, I recommend using keyboard shortcuts. Arrow keys (right/left) are particularly useful, as they move the selection and ensure the selected chord stays in view. You can also check out the [full list of keyboard shortcuts](https://chordseqai.com/wiki/features/keyboard-shortcuts).

**Reply by abulka** [reply]
Ah, I see how zoom works now, you have to deliberately position the mouse at the first chord when zooming in to prevent the first chords being lost out of the view. 

Giving trackpad users a way to "Move the view" is critical - most Mac users don't use mice.

Knowing the keyboard shortcuts is very helpful, thanks. IMO these should be linked from the app's help page, in fact the whole https://chordseqai.com/wiki/introduction website should be accessible from the app's help menu - so much info I didn't know about - all the currently app help page has is the video.

**Reply by PetrIvan** [reply]
You're right, trackpad support wasn't a priority during the initial development (as most DAW users prefer a mouse), although I can see how it would be useful. I'll look into it, if you have any suggestions on how to move the view (such as horizontal scroll), feel free to share!

Regarding the help menu, the wiki actually is available from it, although I agree it could be made more prominent (it is just mentioned in the text with a link).

**Reply by abulka** [reply]
Re "most DAW users prefer a mouse" is this true? Most windows users use a mouse anyway, sure. There is a huge Mac music creative community and most Mac users use trackpads - I've never heard of those Mac DAW users specifically switching to using a mouse for DAW use. Mac DAW users frequently leverage the trackpad for zooming, scrolling, and performing other gestures that enhance workflow, especially in software like Logic Pro X, Ableton Live, and other DAWs.

As for specific suggestions, I'd just go with the standard (used by Ableton, Logic, etc.) two fingers to scroll left and right, one finger click drag and drop. To zoom it would be a two finger pinch. Attached is video of these ideas in Ableton Live.

https://github.com/user-attachments/assets/76d5d453-8b70-4be5-9db1-9a9d00e790c4

Here is a HTML js [jsfiddle demo](https://jsfiddle.net/tcab/mLwfj036/) of the desired behaviour (no drag and drop, just sideways scrolling and pinch zooming, in case that helps.

**Reply by PetrIvan** [reply]
I have to admit that I don't have the specific numbers to show that the mouse is indeed preferred for DAW users, nor have I found any statistics online (just look at [Wikipedia](https://en.m.wikipedia.org/wiki/Pointing_device#:~:text=While%20the%20most%20common%20pointing%20device%20by%20far%20is%20the%20mouse%2C%20many%20more%20devices%20have%20been%20developed.), even they don't provide citations for the claim that the mouse is by far the most common pointing device). From my personal experience, I haven't met many people using trackpads for music production, although I see that it may be more popular in professional circles than I expected. That is why I haven't made the timeline element with trackpad users in mind, it seemed like a minority of potential users.

Thank you for the detailed suggestions though, I'll work on that together with mobile device support (it could be ready around the end of September).

**Reply by PetrIvan** [reply]
Hi @abulka, trackpad gestures should now be available if you want to check it out. Chord reordering is also available, although it turned out to be more difficult to implement than it originally seemed due to many edge cases where it conflicted with other interactions.

**Reply by abulka** [reply]
Thanks so much - now works perfectly and intuitively with a trackpad. Thanks for your effort, I'm sure it wasn't easy.

**Comment by abulka** [comment]
When adding a new chord using the + button or the 'A' hotkey, the chord marked ? should be auto-selected so that the chord choices appear immediately. 

I am forever adding a chord, then having to select the chord to see the chord choices being offered. 

*Aside:*
Correct me if I'm wrong, but in general, adding a chord, selecting it, then clicking on the first chord in the list/matrix would effectively be the main workflow. Repeating this sequence should end up with an AI song? I'm finding the first chord in the matrix not that inspiring and quite repetitive. I'll try playing with the different models a bit more. Ideally a sequence of some jazzy chords would inspire the AI to suggest more jazzy chords but it always seems to suggest plain chords already in the sequence. I'll play with it some more.

**Reply by PetrIvan** [reply]
I have considered that option, but I've decided not to implement auto-selection since it felt finicky while modifying sequences. The left and right arrow keys for selection should make the situation you're describing less exhaustive.

Anyway, always choosing the top prediction is generally not identical to sampling sequences from the AI, as the entire probability distribution should be considered instead. It is like generating sequences from an LLM with the temperature set to zero – it would always produce the same result given the same initial sequence/first chord. It should perhaps be unsurprising that it is repetitive and uninspiring since it is the statistically most likely sequence (although technically greedy search is not guaranteed to find the most probable sequence). Even jazz uses the basic structures most often, which causes the described behavior. Instead, to get a sequence sampled from an AI, you should pick a chord at each step according to its model-predicted probability because even the top prediction might still be weighed at less than 50%.

After some inspection of model behavior and my testing, the choice of the first chord strongly determines the complexity of the entire sequence, as complex sequences usually start with something untrivial. The fact that top predictions are already present in the sequence is also not that surprising, considering that most songs do not use that many chords. I'd suggest choosing something interesting from the beginning if you want to lead the model into predicting something jazzy.

Hope you find that helpful!
